server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1335,14 +1335,20 @@ public DataStream dataStream(String dataStreamName) {
         public Builder dataStreams(Map<String, DataStream> dataStreams, Map<String, DataStreamAlias> dataStreamAliases) {
             previousIndicesLookup = null;
 
+            for (DataStream dataStream : dataStreams.values()) {
+                dataStream.validate(indices::get);
+            }
+
```

Comment:
> That way we don't do validation multiple times if a caller uses this method more than once?

This true. But I wanted to avoid to do this validation if no changes are being made to data streams.
If we move this validation to build() method then this validation is invoked for any change to Metadata (e.g. updating on index setting or template).

If we invoke this validation from the build() method then I think we should perform this validation only when changes to data streams have been made. I think we can do this by adding a boolean flag that is set to true when put() or dataStreams() method is invoked.

Reply:
Do you think we should keep invoking the validation from here? Or should move this to `build()` method behind a boolean flag (so we only invoke this validation when needed)?

Reply:
I think I'm okay with either direction, so I'll leave it up to your preference. If you do leave the validation in the `put*` methods, can you add a comment about why it's here so that future generations will know why we decided that way?